CTF Update Again

AyeZeeBB opened the pull request:
CTF Update
Playtest Ready

Location voting has been added.
New location voting UI.
New location voting victory screen.
New voting music and sound FX.
New respawn menu UI.
New class system via playlist vars.
Legend select in the the respawn menu.
Moved team, flag spawn points into location settings.
Moved deathcam pos, ang into location settings.
Fixed alot of bugs.
Cleaned up code.
Overflow Map Added.

AyeZeeBB committed:
Re CTF Update

AyeZeeBB committed:
Re Update CTF pt2

AyeZeeBB committed:
small update

Fixed end screen menu for maps with only 1 locations ( wasnt displaying correctly )

changed vote menu sound a little bit.
small change to disable buttons after voting ( not needed just cosmetic )

AyeZeeBB committed:
New voting tied ui and some small changes

AyeZeeBB committed:
Cleaned up and some edits

Cleaned up and removed some unused and unneeded code.
Changed sound_accept and sound_focus on voting buttons
Changed sound played when transitioning from winner screen to voting screen

AyeZeeBB committed:
Randomize array

Randomizing the voting maps array just to make it a but more random

IcePixelx committed:
Update ctfvotemenu.menu alignment

IcePixelx commented on vscripts/_remote_functions_mp.gnut:
```diff
@@ -155,10 +155,23 @@ void function RemoteFunctions_Init_Begin()
 			Remote_RegisterUIFunction( "OpenCTFRespawnMenu" )
 			Remote_RegisterUIFunction( "CloseCTFRespawnMenu" )
-			
+			Remote_RegisterUIFunction( "OpenCTFVoteMenu" )
```
Just merge OpenCTFVoteMenu and CloseCTFVoteMenu into 1 Remote call with an integer that then decides behavior via switch case.

This concept can be applied to a fuck ton of function in this gamemode..

IcePixelx commented on vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut:
```diff
@@ -120,6 +137,63 @@ bool function ClientCommand_MIL(entity player, array<string> args)
+
+    int mapid = args[0].tointeger()
+
+    if (mapid == 0)
```
switch case here

IcePixelx replied:
Maybe also check for map name instead?

r-ex replied:
> Maybe also check for map name instead?

no

IcePixelx commented on vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut:
```diff
@@ -204,42 +280,50 @@ void function VotingPhase()
-        CTF.currentmapindex = 0
-
-    if(CTF.setmap)
+    if (CTF.mapvotes1 > 0 || CTF.mapvotes2 > 0 || CTF.mapvotes3 > 0 || CTF.mapvotes4 > 0)
```
This def has to go.

IcePixelx commented on vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut:
```diff
@@ -329,38 +413,241 @@ void function StartRound()
                 }
             }
-            wait 10
+
```
This whole section needs to be re-written cleaner

IcePixelx commented on vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut:
```diff
@@ -1308,40 +1653,43 @@ void function _OnPlayerDied(entity victim, entity attacker, var damageInfo)
-            wait reservedTime
+            if (!CTF.votingtime)
+            {
+                Remote_CallFunction_NonReplay(victim, "ServerCallback_CTF_OpenCTFRespawnMenu", CTF.bubbleCenter, CTF.IMCPoints, CTF.MILITIAPoints, attacker)
```
ServerCallback_CTF_OpenCTFRespawnMenu should also be a ToggleMenu instead.

IcePixelx commented on vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut:
```diff
@@ -25,6 +25,16 @@ global function ServerCallback_CTF_RecaptureFlag
 global function ServerCallback_CTF_EndRecaptureFlag
 global function ServerCallback_CTF_ResetFlagIcons
 global function ServerCallback_CTF_SetPointIconHint
+global function ServerCallback_CTF_OpenCTFVoteMenu
```
Alot of these callbacks can be merged into 1 remote function.

IcePixelx commented on vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut:
```diff
@@ -642,14 +660,180 @@ void function waitrespawn(entity player)
+    RunUIScript( "UpdateVotedLocation", file.locationSettings[mapid].name)
+}
+
+void function UpdateUIVotingLocationTied(int mapid, int done)
```
Switch case here.

IcePixelx commented on vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut:
```diff
@@ -642,14 +660,180 @@ void function waitrespawn(entity player)
+
+void function VoteForMap(int mapid)
+{
+    if(hasvoted)
```
Maybe don't rely on the client to specify if we already voted.

IcePixelx replied:
Handle it server side.

r-ex replied:
it is checked server side as well

IcePixelx commented on vscripts/gamemodes/custom_ctf/sh_gamemode_custom_ctf.nut:
```diff
@@ -349,7 +362,13 @@ vector function GetFlagLocation(LocationSettingsCTF locationSettings, int team)
                 spawnorg = <10684, -18468, -3584>
             break
-        
+        case "Overflow":
```
Switch cases for the team.

IcePixelx commented on vscripts/ui/CTFMenus/menu_ctf_vote.nut:
```diff
@@ -0,0 +1,204 @@
+	AddButtonEventHandler( Hud_GetChild( menu, "MapVote4" ), UIE_CLICK, OnClickMap4 )
+}
+
+void function OnClickMap1( var button )
```
These can all be merged into 1

IcePixelx reviewed (changes requested):
For rexx.

r-ex committed:
ctf cleanup - see desc

changed voting system to use arrays instead of individual vars for each map
removed open/close vote menu functions and create one function that can be used for both

r-ex committed:
fix error after map vote has finished

originally highestVoteId stored the map index instead of the map id so this was an issue

r-ex committed:
formatted ctfvotemenu.menu

r-ex committed:
simplified map vote button callbacks

buttons will now use the "scriptID" field in the ui element to identify themselves within the callback, instead of having separate functions for each button

r-ex committed:
formatted ctfrespawnmenu.menu

r-ex committed:
change WetLands to Wetlands

r-ex committed:
UpdateUIVotingLocationTied switch case

r-ex committed:
punctuation

yes this absolutely needed to be a separate commit

r-ex committed:
fix scriptid issue (uid issue)

r-ex committed:
add purpose comments to some functions

AyeZeeBB committed:
Fixed crash, and script ids

AyeZeeBB committed:
Cleaned up some code, removed unused code, commented code

AyeZeeBB committed:
fix ServerCallback

AyeZeeBB committed:
removed damage taken and received from respawn menu

could never get it to work 100% of the time so ima just remove it for now

AyeZeeBB committed:
moved all set screen callbacks into 1 remote function with switch case

also moved team and enemy flag caps to 1 remote function

AyeZeeBB committed:
make the screens a global enum

AyeZeeBB committed:
moved flag states to global enum

AyeZeeBB committed:
Remove more unneeded code, more comments

AyeZeeBB committed:
New class selection in respawn menu

you can now select between 5 classes when respawning, plan is to hook them up to playlist vars eventually.

also took notes when watching yall clean up the code. this code isn't a bunch of nonsense

AyeZeeBB committed:
remove test clientcommand

AyeZeeBB committed:
Ability to change classes from playlist, see desc for ctf playlist section

https://github.com/I-Am-Zee/ctf_playlist/blob/main/README.md

AyeZeeBB committed:
yay no more shitty switch case

AyeZeeBB committed:
Cleaned up getting playlist vars for classes

also merged enableclassbuttons,  setclassnames, and openctfmenu into 1 function

AyeZeeBB committed:
rename a function.

yes this is a separate commit

AyeZeeBB committed:
player selected class saves across disconnects now

AyeZeeBB committed:
added check for the unlikely chance you die when voting starts

AyeZeeBB committed:
playlist var for filling alt on flag capture

AyeZeeBB committed:
Revamping Respawn ui a little bit, Bottom portion is done. still gotta do the top bit

AyeZeeBB committed:
finished revamping respawn ui

AyeZeeBB committed:
ok it didnt push everything tf

AyeZeeBB committed:
hide change legend

AyeZeeBB committed:
change killed by text color

ok this is the last thing kekw

AyeZeeBB committed:
ok im actually done now ffs

AyeZeeBB committed:
Code Cleanup

AyeZeeBB committed:
small fix to update the clients time left timer if they change resolution

client timer would get out of sync when they change resolution, this fixes that

AyeZeeBB committed:
Huge update, read desc

Big code cleanup.
moved team & flag spawn locations into locations settings.
moved deathcam angles and origin into location settings.
Added legend selection in the respawn menu
Vote screen now shows winning team players and nametags ( like br )
moved around some UI elements

AyeZeeBB committed:
bug fix: wasnt correctly setting clients selected location

AyeZeeBB committed:
playlist var for using default legends ability's instead of classes abilitys

and a very small code cleanup

AyeZeeBB committed:
fix for weapon not being set as the active slot of respawn

AyeZeeBB committed:
freshen up voting menu a little bit

AyeZeeBB committed:
small victory sequence cleanup

AyeZeeBB committed:
finishing touches, see desc

made victory pos registered in location settings.
took deathcam out of LocPairCTF array cuz it didnt need to be in a array.
updated vote screen to show each player on the winning team kills and captures.
small bug fix when the location array = exactly 4.

AyeZeeBB committed:
forgot to remove unused struct

AyeZeeBB committed:
code cleanup read desc

cleaned up some functions,
removed unused functions,
removed and combined some server callbacks,
smoothed out respawn camera,
made respawn timer a playlist var

AyeZeeBB committed:
some more code cleanup, read desc

merged ServerCallback_CTF_RecaptureFlag and ServerCallback_CTF_EndRecaptureFlag

changed some if statements to switch cases

AyeZeeBB committed:
huge server code cleanup

AyeZeeBB committed:
remove change team client command

r-ex committed:
some isvalid checks + formatting

AyeZeeBB committed:
no more low heath hud while in respawn menu

AyeZeeBB committed:
fix for player respawning after legend change

AyeZeeBB committed:
hide compass

AyeZeeBB committed:
crash fix

AyeZeeBB committed:
onplayerconnected, set selected location

AyeZeeBB committed:
bug fixes and some new stuff. read desc

Bug Fixes
Player no longer gets legends abilities on legend change.
Server no longer crashes when flag is being returned and someone picks it up.
Server now sends selected location on player connect.

Changes:
You can now drop the flag with the Inspect key.
You keep your shields as the flag carrier but lose speed boost.

AyeZeeBB committed:
credits update and todo list

AyeZeeBB committed:
maybe fixed kills and captures data for vote screen

AyeZeeBB committed:
more changes, read desc

Added drop flag rui.
Hide flag icon for player holding the flag
Playlist vars for speed boost and take weapons for flag holder

AyeZeeBB committed:
Merge branch 'Mauler125:S3_N1094' into CTF

AyeZeeBB committed:
New main hud score rui, and small bug fixes

AyeZeeBB committed:
some code cleanup

AyeZeeBB committed:
get server time on client connect

r-ex committed:
formatting

AyeZeeBB committed:
lots of bug fixes, see desc

fixed per team highlighting
fixed lots of bugs with returning the flag once its dropped
undermap flag check fix
fixed bug with legend select after changing resolutions
fixed score ui not updating after round change

note: sorry rexx some formating got a little messed up will fix later

AyeZeeBB committed:
Victory Location Improments, also hide stats until i find a good way todo it

AyeZeeBB committed:
fixes and polish, read desc

fully fixed legend select button callbacks

when you die the custom hud ui will no longer show until you are fully respawned

AyeZeeBB committed:
reset weapons and abilities on new round

AyeZeeBB committed:
moved under map z pos into locations settings

AyeZeeBB committed:
formatting

AyeZeeBB committed:
New Locations

KC - Tunnel, Watch Tower

Ash's Redemption

AyeZeeBB committed:
Fixed a IMC spawn at wetlands

AyeZeeBB committed:
small

AyeZeeBB committed:
disable con command for final release

AyeZeeBB committed:
Mirror main hud scorebars

AyeZeeBB commented:
ctf is stable and ready togo.

r-ex committed:
cleanup: formatting

r-ex committed:
more formatting

r-ex committed:
cleanup: formatting

AyeZeeBB committed:
custom rpak, read desc for rpak file

Latest rpak will always be here
https://github.com/I-Am-Zee/ctf_playlist

AyeZeeBB committed:
rpak path rework

AyeZeeBB committed:
better return flag ui,

AyeZeeBB committed:
better rpak paths

AyeZeeBB committed:
more ui polish

AyeZeeBB committed:
small

AyeZeeBB committed:
working on all chat

AyeZeeBB committed:
small cleanup

AyeZeeBB committed:
All chat update for ctf, read desc

I had to edit cl_main_hud to store the text inputed into a global string, was the only way. wont effect any of the normal hud and is only used if someone else wants to implement allchat into their gamemode.

AyeZeeBB committed:
hide chat and overflow fix

AyeZeeBB committed:
Change from gibby bubble to ring

AyeZeeBB committed:
more ring changed

AyeZeeBB committed:
add globalchat playlist var

AyeZeeBB committed:
cleanup

AyeZeeBB committed:
Merge branch 'Mauler125:S3_N1094' into CTF

AyeZeeBB committed:
ring audio fix

AyeZeeBB committed:
ring fix pt.2

AyeZeeBB committed:
comments

AyeZeeBB committed:
Merge remote-tracking branch 'upstream/S3_N1094' into CTF
